demos/device_defender_for_aws/metrics_collector/lwip/metrics_collector.c

```diff
@@ -0,0 +1,209 @@
+ * It can be called with @p pOutConnectionsArray NULL to get the number of
+ * established connections.
+ *
+ * @param[in] pOutConnectionsArray The array to write the established connections
+ * into. This can be NULL, if only the number of established connections is
+ * needed.
+ * @param[in] connectionsArrayLength Length of the pOutConnectionsArray, if it
+ * is not NULL.
+ * @param[out] pOutNumEstablishedConnections Number of the established connections.
+ *
+ * @return #MetricsCollectorSuccess if open TCP ports are successfully obtained;
+ * #MetricsCollectorBadParameter if invalid parameters are passed;
+ */
+    MetricsCollectorStatus_t GetEstablishedConnections( Connection_t * pOutConnectionsArray,
+                                                      uint32_t connectionsArrayLength,
+                                                      uint32_t * pOutNumEstablishedConnections )
+    {
+        MetricsCollectorStatus_t status = MetricsCollectorSuccess;
+        struct tcp_pcb * pCurrPcb = tcp_active_pcbs;
+        uint16_t pcbCnt = 0;
+        Connection_t * pEstablishedConnection;
+
+        if( ( ( pOutConnectionsArray != NULL ) && ( connectionsArrayLength == 0 ) ) ||
+            ( pOutNumEstablishedConnections == NULL ) )
+        {
+            IotLogError( "Invalid parameters. pOutConnectionsArray: 0x%08x,"
+                         " connectionsArrayLength: %u, pOutNumEstablishedConnections: 0x%08x.",
+                         pOutConnectionsArray,
+                         connectionsArrayLength,
+                         pOutNumEstablishedConnections );
+            status = MetricsCollectorBadParameter;
+        }
+
+        if( status == MetricsCollectorSuccess )
+        {
+            LOCK_TCPIP_CORE();
+
+            while( pCurrPcb )
+            {
+                /*  write the ports into pOutConnectionsArray if not NULL */
+                if( ( pOutConnectionsArray != NULL ) && ( pcbCnt < connectionsArrayLength ) )
+                {
+                    /* The output array member to fill. */
+                    pEstablishedConnection = &( pOutConnectionsArray[ pcbCnt ] );
+
+                    ip_addr_copy( pEstablishedConnection->remoteIp, pCurrPcb->remote_ip ); /* network byte order */
+                    ip_addr_copy( pEstablishedConnection->localIp, pCurrPcb->local_ip );   /* network byte order */
+                    pEstablishedConnection->localPort = pCurrPcb->local_port;              /* host byte order */
+                    pEstablishedConnection->remotePort = pCurrPcb->remote_port;            /* host byte order */
+                }
+
+                ++pcbCnt;
+                pCurrPcb = pCurrPcb->next;
+            }
+
+            UNLOCK_TCPIP_CORE();
+        }
+
+        if( ( pOutConnectionsArray != NULL ) && ( pcbCnt > connectionsArrayLength ) )
+        {
+            IotLogWarn( "The connectionsArrayLength is not long engough to store all the established connections" );
```

Comment:
```suggestion
            LogWarn( "The connectionsArrayLength is not long enough to store all the established connections" );
```

Reply:
Have changed by this commit 1c6fcf7cd9de186f0a50788a03f1bf56611d70de